pandas/core/indexes/category.py

```diff
@@ -366,7 +370,7 @@ def argsort(self, *args, **kwargs):
     def _engine(self):
 
         # we are going to look things up with the codes themselves
-        return self._engine_type(lambda: self.codes.astype('i8'), len(self))
```

Comment:
I don't think you need any changes in cython if you simply so: ``.astype('i8', copy=False)``, which is the cause of the perf issue, no?

its still not as nice as actually using type specific hashtables though (which this PR is not addressing)

Reply:
then you have something else going on

Reply:
The numpy docs say about array.astype:

    By default, astype always returns a newly allocated array. If this is set
    to false, and the dtype, order, and subok requirements are satisfied,
    the input array is returned instead of a copy.

So, calling astype(..., copy=False) will only avoid returning a copy when the dtype of codes is int64, i.e. in practice never for Categoricals.

Reply:
so try using ``algos.ensure_int64(values)`` I find this extremely odd that you have to do the if *inside* cython.